runtime/shared_heap.c

```diff
@@ -481,50 +493,118 @@ static intnat pool_sweep(struct caml_heap_state* local, pool** plist,
 
   {
     header_t* p = POOL_FIRST_BLOCK(a, sz);
-    header_t* end = POOL_END(a);
-    mlsize_t wh = wsize_sizeclass[sz];
+    header_t* last_free_block = NULL;
+    const header_t* end = POOL_END(a);
+    const mlsize_t wh = wsize_sizeclass[sz];
     int all_used = 1;
     struct heap_stats* s = &local->stats;
 
-    /* conceptually, this is incremented by [wh] for every iteration
-       below, however we can hoist these increments knowing that [p ==
-       end] on exit from the loop (as asserted) */
+    a->next_obj = 0;
+
     work = end - p;
     do {
-      header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
-      if (hd == 0) {
-        /* already on freelist */
-        all_used = 0;
-      } else if (Has_status_hd(hd, caml_global_heap_state.GARBAGE)) {
+      header_t hd = (header_t)*p;
+
+      if( p + 4096 < end ) {
+        caml_prefetch(p + 4096);
```

Comment:
This `prefetch` logic in the sweeper is new, if I understand correctly. How does it influence the performance comparison between this PR and trunk? Can you add just the prefetching to trunk and see how that works, or remove the prefetching from this PR and see how that works?